sdk/lib/queue/queue.cc

```diff
@@ -0,0 +1,437 @@
+		 */
+		void unlock()
+		{
+			uint32_t value;
+			// Clear the lock bit.
+			value = lockWord.load();
+			// If we're releasing the lock with waiters, wake them up.
+			if (lockWord.exchange(value & ~LockBit) & WaitersBit)
+			{
+				lockWord.notify_all();
+			}
+		}
+	};
+
+	uint32_t counter_load(std::atomic<uint32_t> *counter)
+	{
+		return counter->load() & ~(HighBitFlagLock::reserved_bits());
+	}
+
+	/// Permissions for read-only access to a counter.
+	static constexpr PermissionSet ReadOnly{Permission::Global,
+	                                        Permission::Load};
+	/// Permissions for read-only access to a buffer.
+	static constexpr PermissionSet ReadOnlyCapability{
+	  Permission::Global,
+	  Permission::Load,
+	  Permission::LoadStoreCapability,
+	  Permission::LoadGlobal,
+	  Permission::LoadGlobal};
+	/// Permissions for write-only access to a buffer.
+	static constexpr PermissionSet WriteOnlyCapability{
+	  Permission::Global,
+	  Permission::Store,
+	  Permission::LoadStoreCapability};
+
+} // namespace
+
+struct QueueHandle queue_make_receive_handle(struct QueueHandle handle)
+{
+	Capability buffer   = handle.buffer;
+	Capability producer = handle.producer;
+	buffer.permissions() &= ReadOnlyCapability;
+	producer.permissions() &= ReadOnly;
+	handle.buffer   = buffer;
+	handle.producer = producer;
+	return handle;
+}
+
+struct QueueHandle queue_make_send_handle(struct QueueHandle handle)
+{
+	Capability buffer   = handle.buffer;
+	Capability consumer = handle.consumer;
+	buffer.permissions() &= WriteOnlyCapability;
+	consumer.permissions() &= ReadOnly;
+	handle.buffer   = buffer;
+	handle.consumer = consumer;
+	return handle;
+}
+
+int queue_create(Timeout            *timeout,
+                 struct SObjStruct  *heapCapability,
+                 struct QueueHandle *outQueue,
+                 void              **outAllocation,
+                 size_t              elementSize,
+                 size_t              elementCount)
+{
+	size_t bufferSize;
+	size_t allocSize;
+	bool   overflow =
+	  __builtin_mul_overflow(elementCount, elementSize, &bufferSize);
+	// We must be able to accurately represent the buffer, so round it up to a
+	// representable length.
+	bufferSize = CHERI::representable_length(bufferSize);
+	static constexpr size_t CounterSize = sizeof(uint32_t);
+	// Round up the size to be correctly aligned for the counters at the end
+	// (if necessary) and add two counters worth of space.
+	overflow |= __builtin_add_overflow(
+	  bufferSize,
+	  (2 * CounterSize) + (CounterSize - (bufferSize & (CounterSize - 1))),
+	  &allocSize);
+	if (overflow)
+	{
+		return -EINVAL;
+	}
+	// We need the counters to be able to run to double the queue size without
+	// hitting the high bits.  Error if this is the case.
+	//
+	// This should never be reached: a queue needs to be at least 512 MiB
+	// (assuming one-byte elements) to hit this limit.
+	if (((elementCount * 2) & HighBitFlagLock::reserved_bits()) != 0)
```

Comment:
For maximum paranoia, I think that could be `(elementCount | (elementCount * 2)) & HighBitFlagLoc::reserved_bits()`, lest that `* 2` push the offending bit off the top and badness result.  (The `__builtin_mul_overflow` above won't object to such `elementCount`s with `elementSize == 1`.)